eda/src/autogluon/eda/analysis/shift.py

```diff
@@ -0,0 +1,347 @@
+        self.classifier_class = classifier_class
+        self.split = split
+        self.sample_label = sample_label
+        self.eval_metric = eval_metric
+        self._is_fit = False
+        self._test = None
+        self.test_stat = None
+        self.has_fi = None
+        self.compute_fi = compute_fi
+
+    @staticmethod
+    def _make_source_target_label(data, sample_label):
+        """Turn a source, target pair into a single dataframe with label column"""
+        source, target = data[0].copy(), data[1].copy()
+        source.loc[:,sample_label] = 0
+        target.loc[:,sample_label] = 1
+        data = pd.concat((source, target))
+        return data
+
+    def fit(self, data, **kwargs):
+        """Fit the classifier for predicting if source or target and compute the 2-sample test statistic.
+
+        Parameters
+        ----------
+        data : pd.DataFrame, or tuple
+            either
+            - a dataframe with a label column where 1 = target and 0 = source
+            - a tuple of source dataframe and target dataframe
+        """
+        if isinstance(data, pd.DataFrame):
+            sample_label = self.sample_label
+            assert sample_label in data.columns, "sample_label needs to be a column of data"
+            assert self.split, "sample_label requires the split parameter"
+            data = data.copy() # makes a copy
+        else:
+            assert len(data) == 2, "Data needs to be tuple/list of (source, target) if sample_label is None"
+            data = self._make_source_target_label(data, self.sample_label) # makes a copy
+        if data.index.has_duplicates:
+            data.index = pd.RangeIndex(data.shape[0])
+        train = data.sample(frac=self.split)
+        test = data.drop(train.index)
+        self.classifier.fit(train, **kwargs)
+        yhat = self.classifier.predict(test)
+        self.test_stat = self.eval_metric(test[self.sample_label], yhat)
+        self.has_fi = (getattr(self.classifier, "feature_importance", None) is not None)
+        if self.has_fi and self.compute_fi:
+            self._test = test  # for feature importance
+        self._is_fit = True
+
+    @post_fit
+    def _pvalue_half_permutation(self,
+                                 num_permutations=1000):
+        """The half permutation method for computing p-values.
+        See Section 9.1 of https://arxiv.org/pdf/1602.02210.pdf
+        """
+        perm_stats = [self.test_stat]
+        yhat = self.classifier.predict(self._test)
+        for i in range(num_permutations):
+            perm_yhat = np.random.permutation(yhat)
+            perm_test_stat = self.eval_metric(
+                self._test[self.sample_label],
+                perm_yhat
+            )
+            perm_stats.append(perm_test_stat)
+        pval = (self.test_stat <= np.array(perm_stats)).mean()
+        return pval
+
+    @post_fit
+    def pvalue(self,
+               num_permutations: int=1000):
```

Comment:
ditto: do we need `1000` rounds to get the conclusion?

Reply:
well, we can't resolve a very low pvalue otherwise, think of 1000 permutations at 0.01 level to mean 10 sample above the threshold.